### PR TITLE
Fix trimming issue with extension data

### LINKED ISSUE #52662: Using JSON Source Gen still roots all JsonConverters
Using System.Text.Json source generation in my application isn't providing for size reductions in the app because all the JsonConverters are still being rooted in the app.

### Repro

```C#
using System;
using System.Text.Json;
using System.Text.Json.Serialization;
using ConsoleApp13;
using ConsoleApp13.JsonSourceGeneration;

[assembly: JsonSerializable(typeof(WeatherForecast))]

namespace ConsoleApp13
{
    class Program
    {
        static void Main(string[] args)
        {
            string f = @"{""TemperatureC"":0,""Summary"":""hi""}";

            var forecast = JsonSerializer.Deserialize(f, new JsonContext().WeatherForecast);
            Console.WriteLine(forecast);
        }
    }

    public record WeatherForecast
    {
        public DateTime Date { get; set; }

        public int TemperatureC { get; set; }

        public string Summary { get; set; }

        public int TemperatureF => 32 + (int)(TemperatureC / 0.5556);
    }
}
```

`dotnet publish -r win-x64 -c Release -p:PublishTrimmed=true -p:DebuggerSupport=false`

View the outputted assemblies, and notice that `System.Text.Json.dll` is ~289KB, it still has all its JsonConverters in it, and `System.Collections.Immutable.dll` is in my app, even though I don't use it.

### Analysis

From looking at the results, the reason everything is still rooted is because of the following call graph:

image

This is what is rooting `ObjectConverter`. Then inside `ObjectConverter.Read`, it uses `JsonNodeConverter.Instance`:

https://github.com/dotnet/runtime/blob/b4f4418d56d32c3a1677b5651899a112d7e02226/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/ObjectConverter.cs#L15-L22

This brings in all of the `JsonNode` hierarchy, including `JsonValue<T>`, which overrides `ToString()` to call `JsonSerializer.Serialize` without source gen information:

https://github.com/dotnet/runtime/blob/1410c12ef8068b06518ec99b6cc259dd0c448892/src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueOfT.cs#L89

That this point, since `JsonSerializer.Serialize` is called without source gen information, it now roots all JsonConverters, and the size of the app grows by a lot.

cc @layomia @steveharter @marek-safar @vitek-karas @CoffeeFlux

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/52662

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs [hunk 1]
```diff
@@ -389,7 +389,7 @@ private bool NumberHandingIsApplicable()
                     else
                     {
                         // Slower path for non-generic types that implement IDictionary<string, object>.
-                        converter = JsonMetadataServices.ObjectConverter;
+                        converter = Options.GetConverterInternal(JsonTypeInfo.ObjectType);
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs:392]: Any size savings if we do the same on L405 or is `JsonElement`/`JsonElementConverter` still rooted elsewhere?

**Review by layomia — APPROVED** [review]
beautiful

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs (hunk 1 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs]: There is potential savings here by not rooting; pushed another commit that removes the reference to `JsonMetadataServices`.

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/ExtensionDataTests.cs [hunk 2]
```diff
@@ -1239,7 +1239,7 @@ public static void CustomJsonElementConverterInExtensionProperty()
 
             ClassWithExtensionPropertyAsJsonElement obj = JsonSerializer.Deserialize<ClassWithExtensionPropertyAsJsonElement>(Json, options);
             JsonElement overflowProp = obj.MyOverflow["hello"];
-            Assert.Equal(JsonValueKind.String, overflowProp.ValueKind);
+            Assert.Equal(JsonValueKind.Undefined, overflowProp.ValueKind);
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/ExtensionDataTests.cs:1242]: What causes the change here? Do we need to document a breaking change?

Comment (steveharter) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/ExtensionDataTests.cs:1242]: This test change is correct. There was a bug previously where the custom converter was only used on write, not read.

**Pull request merged**